c7n/resources/vpc.py

```diff
@@ -2710,3 +2710,75 @@ class resource_type(query.TypeInfo):
         arn_type = 'traffic-mirror-target'
         universal_taggable = object()
         id_prefix = 'tmt-'
+
+
+@RouteTable.filter_registry.register('cross-az-nat-gateway-route')
+class CrossAZRouteTable(Filter):
+    """Filter route-tables to find those with routes which send traffic
+    from a subnet in an az to a nat gateway in a different az.
+
+    This filter is useful for cost optimization, resiliency, and
+    performance use-cases, where we don't want network traffic to
+    cross from one availability zone (AZ) to another AZ.
+
+    :Example:
+    .. code-block:: yaml
+            policies:
+              - name: cross-az-nat-gateway-traffic
+                resource: aws.route-table
+                filters:
+                    - type: cross-az-nat-gateway-route
+                actions:
+                  - notify
+
+    """
+    schema = type_schema('cross-az-nat-gateway-route')
+    permissions = ("ec2:DescribeRouteTables", "ec2:DescribeNatGateways", "ec2:DescribeSubnets")
+
+
+    def resolve_subnets(self, resource, subnets):
+        subnet_ids = set()
+        for association in resource['Associations']:
+            if association.get('Main'):
+                subnet_ids.update({
+                    s['SubnetId'] for s in subnets if s['VpcId'] == resource['VpcId']})
```

Comment:
@kapilt does this actually do what we want? If we're dealing with the main route table, then this code will return ALL subnets (in the same VPC) as associated subnets. I don't think that's what we want. For main route table, we should only return subnets (in the same VPC) that are not explicitly associated in other route tables right?

Reply:
yup, good catch